sandbox/apparmor/notify/listener/listener.go

```diff
@@ -464,12 +460,17 @@ func (l *Listener) decodeAndDispatchRequest(buf []byte) error {
 		}
 
 		// Before sending the request, check if it should instead be queued.
-		if l.pendingCount > 0 && !msg.Resent() {
-			// Not a previously-sent pending message, so it's from the
-			// kernel's ready queue, not the pending queue. Queue it up to
-			// be sent once all pending messages have been re-received.
-			l.readyQueue = append(l.readyQueue, req)
-			continue
+		if !msg.Resent() {
+			// This is a new message, not one we're re-receiving, so if the
+			// listener is not yet ready, queue it up for later instead.
+			l.pendingMu.Lock()
+			if l.pendingCount > 0 {
+				l.readyQueue = append(l.readyQueue, req)
+				l.pendingMu.Unlock()
+				continue
+			}
+			l.pendingMu.Unlock()
```

Comment:
I recommend creating small helpers around anything that touches pendingMu or pendingCount, this could be for example appendToReadyQueue and return a boolean whether it did that or not, etc